Implement lexer and parser

theY4Kman opened the pull request:
# Brief

I spent some real time building out a SourcePawn grammar. It's not yet at the point of parsing full sources, yet, but it's a shitload better than whatever I checked in years ago...

Closes #3
Refs #5


# Parser

At the moment, the grammar supports:
 - Global declarations/initializations (without pre-1.7 support, or dimensions, I believe)
 - Function definitions
 - Local declarations/initializations (sans pre-1.7 support)
 - Integer, float, hex, binary, string, and char literals
 - Full expressions (arithmetic, assignment, array subscripts, paren expressions — no method calls, yet, though)
 - if/else, switch statements
 - for, while, do/while loops
	- continue/break statements
 - Struct literals (e.g. `public Plugin myinfo = {  }`)
 - native & forward declarations
 - enums
 - typedefs
 - typesets
 - Object method calls (e.g. `myHandle.Method()`)
 - Array item assignment (e.g. `myArray[1] = 2`)
 - `_` keyword (e.g. `CallAThing("test", _, 2)`)
 - Expressions in global decl dims, to allow for preprocessor defines (e.g. `bool g_isAdmin[MAXPLAYERS+1]`)

It's missing:
 - `new`/`decl` operators
 - Labels/tags in various places
 - Parsing of preprocessor directives
 - ... and I'm sure plenty of other stuff I'm forgetting. It's been a while.


# Grammar

The JFlex lexer definition was updated to properly handle block comments.

Also, the lexer now returns a different token for doc comments (`/** */`), so they may be highlighted differently.

It appears the lexer must also be updated to parse string and character literals correctly. I think they're regex-based right now, leading to weird shit like this:

![image](https://user-images.githubusercontent.com/33840/106584522-4c5f5700-6514-11eb-8fca-34ef17abc36f.png)


# Syntax highlighting

The syntax highlighter has been adjusted to provide a wee bit better experience.

![image](https://user-images.githubusercontent.com/33840/106583560-3a30e900-6513-11eb-8856-e84c8c0db1a4.png)

theY4Kman committed:
feat(parse): rewrite grammar/lexer def from ground up

theY4Kman commented:
Also, I wondered a couple things

 - When playing around with the grammar, I ran into some staleness issues: if I deleted or renamed some symbol, the corresponding element interface/implementation file still remained, and would cause errors when compiling to run the IDE. I changed my workflow to "delete the contents of the `gen` folder, then hit runIde". 
 
   I wonder if there's a better way to do that – perhaps some option in GrammarKit to do a more thorough cleaning; perhaps explicitly adding the `gen` cleaning to the `runIde` gradle task; or perhaps a separate gradle task – maybe `runIdeClean` or `runIdeWithFreshGrammar`

 - This PR is enormous due to the changes in `gen`. Is it customary to include the generated artifacts in version control?

---

On the grammar: at the moment, I'm only concerned about breadth and correctness of parsability – I just want it to be able to parse the full set of includes and stock plugins. I'm not yet aiming for ease of tree manipulation once parsed – i.e. the ability to do shit with the parsed elements, like inspections or type-checking or references. I wonder how the grammar might change to accommodate that use case. I love all the goodies I find in the Python plugin's elements (e.g. [`PyParameterList`](https://github.com/JetBrains/intellij-community/blob/43e194628e54fa708a0e6fbe7c193bc358a99ba7/python/python-psi-api/src/com/jetbrains/python/psi/PyParameterList.java)), and am excited to see similar utility in idea-sp.

Hexer10 commented on src/main/grammar/SourcePawn_test.bnf:
```diff
@@ -0,0 +1,197 @@
+storage_class ::= "public" | "static" | "const" | "stock"
+
+type_expr ::= (builtin_type | symbol) new_dims?
+builtin_type ::= "void"
```
Here is missing the `any` type. Also I'm not sure how the `Handle` type should be handled, sourcepawn declares it as an `enum` but it is actually treated as a type.

theY4Kman replied:
Ah, word. I skipped doing tags and some other pre-1.7 stuff — I presume `any` is only used as a tag (i.e. `new any:myvar = 123`) and not a fancy new type (e.g. `any myvar = 123;`). Lemme add all that in.

Thanks for taking a look here :)

Hexer10 replied:
Yeah, `any` accepts any type, here are some valid examples:
```
    any a = 'a';
    any b = 1;
    any c = 1.1;
    any d = new ArrayList(); // Or any other handle
    any e[] = "hello"; // `any x = ""` is not valid


 ```

theY4Kman replied:
Oh, shit, word – like a C++ `auto`. I'll add her in.

I'll push up my latest changes later tonight. I believe I've since added:
 - Dimensions for declarations
 - The `view_as` operator
 - methodmaps, including properties (though, I missed inheritance and destructors)
 - native & forward declarations
 - I think I added function calls

Still missing:
 - typedefs
 - typesets
 - enum structs
 - new/delete expressions
 - Support for function calls with named parameters
 - Array literals (`int butts[] = { 1, 2, 3 }`), I believe
 - ... including ellipsis support in array literals
 - Also, ellipsis support for concatenating string literals
 - And come to think of it, likely `\` support for string literal concat, too

Hexer10 replied:
The `sizeof` operator is missing as well, it should be noted that both these annotations are valid:
```
sizeof(var); // Format(var, sizeof(var), ...);
sizeof var // Format(var, sizeof var, ...);
```

Also the `any` type iirc is not like C++ `auto` since you can reassign the variable to any type you want:
```
any x = 1;
x = 2.1;
x = '';
x = new ArrayList();
```

Hexer10 commented on src/main/grammar/SourcePawn.flex:
```diff
@@ -111,11 +131,16 @@ SYMBOL=([@_a-zA-Z][@_a-zA-Z0-9]+|[a-zA-Z][@_a-zA-Z0-9]*)
   "decl"                      { return DECL_KEYWORD; }
   "enum"                      { return ENUM_KEYWORD; }
   "struct"                    { return STRUCT_KEYWORD; }
+  "void"                      { return VOID_TYPE; }
```
Same as before.
> Here is missing the any type. Also I'm not sure how the Handle type should be handled, sourcepawn declares it as an enum but it is actually treated as a type.

Hexer10 commented:
To purge the old files when generating the grammar it should be enough to set `purgeOldFiles = true` in the [gradle task](https://github.com/JetBrains/gradle-grammar-kit-plugin#tasks).

I'm sure how common it is to push the generate files into version control, tough it might be better to gitignore them review PRs more easily.

theY4Kman commented:
> I'm sure how common it is to push the generate files into version control, tough it might be better to gitignore them review PRs more easily.

FYI: I went ahead and killed the generated files w/ #10

theY4Kman committed:
chore(gradle): add purgeOldFiles to lexer/parser gen tasks

I'm not sure it works as intended, though

theY4Kman committed:
fix(parse): allow multiple top-level statements

Whoops!

theY4Kman committed:
chore(parse): add grammar support for methodmaps, enums, native/forward decls, struct literals (e.g. Plugin myinfo), view_as<> operator, using statement (used once in handles.inc), new/old dimensions support in var decls

theY4Kman committed:
fix(lex): allow newlines in block comments

Whooops!

theY4Kman committed:
chore(lex): parse doc comments (/** */) separately from block comments (/* */)

This is done to highlight them separately

theY4Kman committed:
feat(editor): expand syntax highlighting

theY4Kman committed:
feat(parse): add any as type keyword

I think this may end up getting reverted or implemented another way, because 'any' is _not_ a reserved word; it is apparently perfectly valid to write 'any any = 1;' — and the same is true for void, int, float, char, and bool.

theY4Kman committed:
feat(parse): add support for dimensioned var decls (e.g. `int stuff[2]` or `int stuff[] = ...`)

theY4Kman committed:
feat(parse): add support for array literals (`int things[] = { 1, 2, 3, ... }`), including ellipsis

theY4Kman committed:
feat(parse): add null as keyword

theY4Kman committed:
fix(parse): remove ambiguous pin causing overly-greedy local_decl rule

theY4Kman committed:
feat(parse): add support for new expr, function calls

theY4Kman committed:
fix(parse): add support for inheritance in methodmaps

theY4Kman committed:
fix(parse): allow trailing semicolon after methodmap decls

Hexer10 commented on src/main/grammar/SourcePawn.flex:
```diff
@@ -5,26 +5,37 @@ import static org.idea_sp.psi.SourcePawnTypes.*;
 LINE_COMMENT="//"[^\r\n]*
-BLOCK_COMMENT="/"\*(.|\n)*?\*"/"
+BLOCK_COMMENT=[\s\S]
+DOC_COMMENT=[\s\S\n]
```
Isn't `[^]` the same as `[\s\S\n]`? Also `\s` should already include `\n`

theY4Kman replied:
I don't believe I had a good grasp of the purpose of these definitions when I wrote them. I think I believed they became token or element types, so I thought it was necessary to define them in order to `return` them for use in the parser... but that appears untrue. Instead, the grammar's `tokens` definition seems to be the source of truth.

The regexes defined for `BLOCK_COMMENT` and `DOC_COMMENT` are never used, since we just look for their starting sequences and consume all tokens till their ending sequences. (Though, I guess the token consumption _does_ use `[\s\S\n]`.) Those definitions should be removed for clarity.

I had assumed `[\s\S]` would include all characters; however, I found that after adding `\n` to the character group, it properly parsed multiline comments. I wasn't really scientific enough about it to say whether one caused the other, and not something else I'd done... Regardless `[^]` is definitely much more succinct. I wish _I'd_ thought of that :P (Or looked down to the `BAD_CHARACTER` rule)

theY4Kman commented:
> I think this may end up getting reverted or implemented another way, because 'any' is _not_ a reserved word; it is apparently perfectly valid to write 'any any = 1;' — and the same is true for void, int, float, char, and bool.

I wrote this in a commit message, and just wanted to clarify for posterity: this is not wholly true. It _is_ true that `any any = 1;` is valid, but the remaining examples (`bool bool = true`, etc) are _not_ valid. I had been using the [sourcemod.net compiler](https://sourcemod.net/compiler.php) to test things, and had apparently written my test snippets in a way that didn't reveal the truth (read: shitty).

theY4Kman commented:
It seems wise to jot down other things I've encountered while continuing to flesh the lexer/parser out. There have been times I've wondered about lines in the lexer def or grammar, and really wish I'd've known where they came from or why they were chosen... so, I may be unable to answer those questions, but I can at least pay it forward.


### String Lexing/Parsing

While toying with a more tuned scheme for lexing string literals, I realized how the lexer for an IDE would differ from that of the compiler: while the compiler ultimately only cares about the resulting string of characters the literal produces, the IDE must additionally care about things like tokenizing escape sequences, so they may be syntax-highlighted differently.

As the string lexing is furthered, I imagine there being two tokens comprising any string literal: an escape sequence token, and a string character token (likely a char class including the double-quotes). Then, the parser would put together these tokens to form the _actual_ bedrock `string_literal` element.


### Robustness of Error Handling

As it stands, the grammar is, bluntly, quite shit at recovering from errors. Though pinning does seem to avoid the earlier situation where the first parsing error would cause the remainder of the file to fall on its face, I still find cases where an error in one statement in a function causes all the remaining statements to get bungled.

For instance, before I added a rule for the `new` operator, I encountered a situation like this:
```sp
bool test = true; // <-- red squigglies at/after semicolon
int[] arr = new int[20];
```

Even though the grammar could very well parse the var decl part (`int[] arr`), and the initializer was pinned to the equals sign, the error was regardless attributed to `int` – the token following the last properly-parsed token (the semicolon). Ideally, the error and red squigglies would be applied entirely to the then-unsupported `new int[20]` expression, with the rest of the function body remaining unaffected.

I believe this sort of issue is what [`recoverWhile`](https://github.com/JetBrains/Grammar-Kit/blob/master/HOWTO.md#22-using-recoverwhile-attribute) is intended to address, but my attempts to apply it have yielded different results than I intend.

My resolved intention is to complete the parser to a full grammar capable of handling the entire suite of includes and stock plugins, but without proper error handling, actual writing of SourcePawn will be janky: there will always be times when a statement is yet to be fully-formed... the IDE should not barf in these cases.


### Preprocessor Directives

Currently, preprocessor directives (`#if`, `#include`, `#pragma`, etc) are lexed as comments, and summarily ignored. I do not see it remaining this way. I see these directives being lexed as tokens, likely using purpose-built lexer states to handle their fine details.

For instance, `#include` will likely evolve to a state to lex `"./relative/paths"` and `<sdk/paths>` as separate tokens. `#pragma` may lex known compiler options as keyword tokens. `#if` must lex its expressions, so they may be syntax-highlighted and parsed for verification. `#define` as well... that's a whole bag of cans of worms.

theY4Kman commented:
Well, I got the grammar to the point where it can parse the entire `include/` directory without error!

I believe all that's missing is pre-1.7 support: labels/tags (for vars, args, return values, etc), labels/tags as reinterpret_casts, new/decl operators, and functags/funcenums. Oh, shit, and I don't think I added `enum struct` support for 1.7+ (well, according to the docs, 1.10+).

theY4Kman commented:
I'm just gonna push a WIP commit for now, and I'll be back in a couple hours to split it off into multiple commits. Not that it much matters :P

theY4Kman commented on src/main/grammar/SourcePawn.bnf:
```diff
@@ -3,6 +3,9 @@
   parserClass="org.idea_sp.parser.SourcePawnParser"
 
   extends="com.intellij.extapi.psi.ASTWrapperPsiElement"
+  extends(".*_expr")=expr
```
By making all expression rules inherit from the top-level expression rule, GrammarKit apparently uses a different method of parsing (Pratt?) which eliminates left-recursion — something the regularly-scheduled parser won't handle.

theY4Kman commented on src/main/grammar/SourcePawn.bnf:
```diff
@@ -93,83 +99,295 @@
 
     string_literal="regexp:\"(\\.|[^\"])*\""
-    character_literal="regexp:'(\\.|[^\"])'"
+    character_literal="regexp:'(\\.|[^'])'"
```
This was a mistake in the initial commits from years ago (also reflected in the lexer def).

theY4Kman commented on src/main/grammar/SourcePawn.bnf:
```diff
@@ -93,83 +99,295 @@
-//    symbol="regexp:[A-Za-z_]([A-Za-z0-9_@]*)"
     symbol="regexp:([@_a-zA-Z][@_a-zA-Z0-9]+|[a-zA-Z][@_a-zA-Z0-9]*)"
+
+    EOL="regexp:\r|\n|\r\n"
```
I'd been toying with the inclusion of end-of-line tokens — to help parse preprocessor directives and the such. I'm not sure it makes much sense to do so, however.

theY4Kman commented on src/main/grammar/SourcePawn.bnf:
```diff
@@ -93,83 +99,295 @@
+post_increment_expr ::= expr PLUSPLUS { pin = 2 }
+post_decrement_expr ::= expr MINUSMINUS { pin = 2 }
+subscript_expr ::= expr subscript { pin = 2 }
+// XXX: lack of expression used for sizeof expr. is there a better way?
```
By this, I mean the measurement of a type nested in an array, like this:
```sourcepawn
sizeof(my_array[])
```

theY4Kman commented on src/main/grammar/SourcePawn.flex:
```diff
@@ -5,125 +5,191 @@ import static org.idea_sp.psi.SourcePawnTypes.*;
+}
+
+<P_COMMENT> {
+  {BLOCK_COMMENT_DOC_START}    { tmp = elementStart;
+                                 elementStart = zzCurrentPos;
+                                 commentType = DOC_COMMENT;
+                                 yybegin(COMMENT);
+                                 zzMarkedPos -= yylength();
+                                 zzStartRead = tmp;
+                                 return PREPROCESSOR_COMMENT; }
```
all this business of screwing with lexer positionings is pretty darned gross. there is likely a much simpler and prettier way of doing this. At the very least, this hair could be extracted out to a method.

theY4Kman commented on src/main/kotlin/org/idea_sp/SourcePawnColorSettingsPage.kt:
```diff
@@ -10,10 +11,30 @@ import javax.swing.Icon
-            AttributesDescriptor("Literal", SourcePawnSyntaxHighlighter.LITERAL),
-            AttributesDescriptor("Reserved", SourcePawnSyntaxHighlighter.RESERVED),
-            AttributesDescriptor("Comment", SourcePawnSyntaxHighlighter.COMMENT),
+            AttributesDescriptor(
+                "Operator", TextAttributesKey.createTextAttributesKey(
+                    "SOURCE_PAWN_OPERATOR",
+                    DefaultLanguageHighlighterColors.OPERATION_SIGN
+                )
+            ),
+            AttributesDescriptor(
+                "Literal", TextAttributesKey.createTextAttributesKey(
+                    "SOURCE_PAWN_CONSTANT",
+                    DefaultLanguageHighlighterColors.CONSTANT
+                )
+            ),
+            AttributesDescriptor(
+                "Reserved", TextAttributesKey.createTextAttributesKey(
+                    "SOURCE_PAWN_RESERVED",
+                    DefaultLanguageHighlighterColors.KEYWORD
+                )
+            ),
+            AttributesDescriptor(
+                "Comment", TextAttributesKey.createTextAttributesKey(
+                    "SOURCE_PAWN_COMMENT",
+                    DefaultLanguageHighlighterColors.LINE_COMMENT
+                )
+            ),
```
Whoops, this was a mistake — I meant to inline some variables in another file.

Also, this settings page will need some changing after the syntax-highlighter is all settled.

theY4Kman commented on src/main/kotlin/org/idea_sp/SourcePawnTokenTypeSets.kt:
```diff
@@ -40,8 +42,62 @@ interface SourcePawnTokenTypeSets {
+            SourcePawnTypes.PROPERTY_KEYWORD,
+            SourcePawnTypes.VIEW_AS_KEYWORD
+        )
+        val OPERATORS = TokenSet.create(
```
These lists are by no means complete, as well. As I've fiddled with the parser, I've neglected updates to this file. As well, no sort of grand planning came into the organization here. It could sure do with some — organic or otherwise.

theY4Kman commented on src/main/resources/META-INF/plugin.xml:
```diff
@@ -12,6 +12,9 @@
     <extensions defaultExtensionNs="com.intellij">
         <!-- TODO: Replace fileTypeFactory with fileType -->
         <fileTypeFactory implementation="org.idea_sp.SourcePawnFileTypeFactory"/>
+<!--        <fileType name="SourcePawn file" language="SourcePawn"-->
+<!--                  implementationClass="org.idea_sp.SourcePawnFileType"-->
+<!--                  extensions="sp,inc" />-->
```
I was trying :P

theY4Kman committed:
feat(parse): implement all 1.7 features

This includes:
 - Identifiers with dots or double-colons in them (e.g. `g_cvar.AddStuff()` or `ConVar::AddStuff`)
 - Underscore as a reserved word (NOTE: underscore as a label/tag not yet implemented)
 - true/false as reserved words
 - typesets/typedefs
 - The function type
 - struct declarations
 - Support for ellipsis varargs in func decls
 - Array element assignment is now supported
 - The sizeof operator

Additionally, some other miscellaneous things include:
 - Block comments are now supported in preprocessor comments
 - A faulty character literal regex has been fixed.
 - Added built-in types and typedef/typeset tokens to syntax highlighter
 - More stringent lexing of number literals, to avoid arithmetic
   expressions such as `MAXPLAYERS+1` from being parsed as
   `SYMBOL INTEGER_LITERAL`, instead of `SYMBOL OP INTEGER_LITERAL`
 - Some rules allow trailing commas (such as typedef/typeset decls).
   These are now supported
 - The `float()` native declaration for older versions of SP is parsed
   correctly. (NOTE: this method is language built-in in newer versions.)
 - break/continue were only allowed directly underneath a loop (and not,
   say, within a child `if`). This restriction has been lifted.
 - The `<=` and `>=` operators are now properly supported (they were
   missing from the lexer altogether)

theY4Kman commented:
Okay, pre-1.7 support is not finished, but I'm going to overlook that for now. I cut a lot of corners in various places, but I think the best route to cleaning them up is to move on to editor stuff, like references. Only then, I believe, will I be able to see where to improve.

So, I'm gonna merge this guy and see if I can get Find Usages and Go To Declaration to work (at least within a single file)
